iree/turbine/kernel/wave/utils.py

```diff
@@ -1043,3 +1044,208 @@ def evaluate_with_assumptions(constraints: list[Constraint], expr: IndexExpr) ->
     if isinstance(result, sympy.logic.boolalg.BooleanAtom):
         return False
     return True if any([result.equals(x) for x in facts]) else None
+
+
+def _get_start_index(i: IndexSequence | IndexExpr) -> IndexExpr:
+    if isinstance(i, IndexSequence):
+        i = i.start
+
+    return i
+
+
+def _get_start_indices(
+    src_indices: dict[IndexExpr, IndexSequence | IndexExpr]
+) -> list[IndexExpr]:
+    start_indices = []
+    for dim_indexing in src_indices:
+        i = _get_start_index(src_indices[dim_indexing])
+        start_indices.append(i)
+
+    return start_indices
+
+
+def _simplify_sympy_expr(expr: IndexExpr) -> IndexExpr:
+    """Apply custom sympy simplifications"""
+
+    def check_mul(mul):
+        ret = None
+        for arg in mul.args:
+            if arg.is_number:
+                if ret is not None:
+                    return None
+
+                ret = arg
+                continue
+
+            if not isinstance(arg, (sympy.floor, sympy.Mod)):
+                return None
+
+        return ret
+
+    def transform_mod(expr):
+        """Move constant outside of Mod expr
+
+        Example:
+        (floor(a) * 4 + 3) % 16 -> (floor(a) * 4) % 16 + 3
```

Comment:
This transformation isn't true even for positive numbers.
See this example:
```
(2 + 2) % 3 = 1 which is not equal to (2 % 3) + (2 % 3)
```
However, what you can say is
```
(a + b) % M = ((a % M) + (b % M)) % M
```

Reply:
I'm matching expr `(floor(x) * a + b) % c` and checking `c % a == 0` and `b < a`

Reply:
Makes sense, but I'm not sure that is enough for correctness. What you want is
```
(floor(x) * a + b) % c == (floor(x) * a) % c + b % c
```
which is only true when either b is a multiple of c, or 
```
 (floor(x) * a) % c + (b % c) < c
```